runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java

```diff
@@ -117,15 +120,21 @@ DataflowPipelineJob run(Pipeline pipeline, DataflowRunner runner) {
           public Optional<Boolean> call() throws Exception {
             try {
               for (;;) {
-                Optional<Boolean> result = checkForSuccess(job);
-                if (result.isPresent()) {
+                JobMetrics metrics = getJobMetrics(job);
+                Optional<Boolean> result = checkForSuccess(job, metrics);
+                if (result.isPresent() && (!result.get() || atMaxWatermark(job, metrics))) {
```

Comment:
Since you are changing here, could you give variable "result" a better name?

For example, !pAssertPassed.get() or !success.get(). Both are better than !result.get(), because result doesn't give any context.